File: python_bindings/correctness/tuple_select.py
import halide as hl
import numpy as np


def test_tuple_select():
    x = hl.Var('x')
    y = hl.Var('y')

    # ternary tuple_select with Expr condition
    f = hl.Func('f')
    f[x, y] = hl.tuple_select(x + y < 30, (x, y), (x-1, y-2))

    a, b = f.realize([200, 200])
    for xx in range(a.height()):
        for yy in range(a.width()):
            correct_a = xx if xx + yy < 30 else xx-1
            correct_b = yy if xx + yy < 30 else yy-2
            assert a[xx, yy] == correct_a
            assert b[xx, yy] == correct_b

    # ternary tuple_select with Tuple condition
    f = hl.Func('f')
    f[x, y] = hl.tuple_select((x < 30, y < 30), (x, y), (x-1, y-2))

    a, b = f.realize([200, 200])
    for xx in range(a.height()):
        for yy in range(a.width()):
            correct_a = xx if xx < 30 else xx-1
            correct_b = yy if yy < 30 else yy-2
            assert a[xx, yy] == correct_a
            assert b[xx, yy] == correct_b

    # multiway tuple_select with Expr condition
    f = hl.Func('f')
    f[x, y] = hl.tuple_select(x + y < 30,  (x, y),
                              x + y < 100, (x-1, y-2),
                                           (x-100, y-200))

    a, b = f.realize([200, 200])
    for xx in range(a.height()):
        for yy in range(a.width()):
            correct_a = xx if xx + yy < 30 else xx-1 if xx + yy < 100 else xx - 100
            correct_b = yy if xx + yy < 30 else yy-2 if xx + yy < 100 else yy - 200
            assert a[xx, yy] == correct_a
            assert b[xx, yy] == correct_b

    # multiway tuple_select with Tuple condition
    f = hl.Func('f')
    f[x, y] = hl.tuple_select((x < 30, y < 30),   (x, y),
                              (x < 100, y < 100), (x-1, y-2),
                                                  (x-100, y-200))

    a, b = f.realize([200, 200])
    for xx in range(a.height()):
        for yy in range(a.width()):
            correct_a = xx if xx < 30 else xx-1 if xx < 100 else xx - 100
            correct_b = yy if yy < 30 else yy-2 if yy < 100 else yy - 200
            assert a[xx, yy] == correct_a
            assert b[xx, yy] == correct_b

    # Failure case: mixing Expr and Tuple in multiway
    try:
        f = hl.Func('f')
        f[x, y] = hl.tuple_select((x < 30, y < 30), (x, y),
                                   x + y < 100,     (x-1, y-2),
                                                    (x-100, y-200))
    except hl.HalideError as e:
        assert 'tuple_select() may not mix Expr and Tuple for the condition elements.' in str(e)
    else:
        assert False, 'Did not see expected exception!'

    # Failure case: Tuples of mixed sizes
    try:
        f = hl.Func('f')
        f[x, y] = hl.tuple_select((x < 30, y < 30), (x, y, 0), (1, 2, 3, 4))
    except hl.HalideError as e:
        assert 'tuple_select() requires all Tuples to have identical sizes' in str(e)
    else:
        assert False, 'Did not see expected exception!'


if __name__ == "__main__":
    test_tuple_select()

File: test/correctness/tuple_select.cpp
#include "Halide.h"
#include <gtest/gtest.h>

using namespace Halide;
using namespace Halide::Internal;

TEST(TupleSelect, TernaryWithExprCondition) {
    // ternary select with Expr condition
    Var x("x"), y("y");
    Func f("f");

    f(x, y) = select(x + y < 30, Tuple(x, y), Tuple(x - 1, y - 2));

    Realization result = f.realize({200, 200});
    Buffer<int> a = result[0], b = result[1];
    for (int y = 0; y < a.height(); y++) {
        for (int x = 0; x < a.width(); x++) {
            int correct_a = (x + y < 30) ? x : x - 1;
            int correct_b = (x + y < 30) ? y : y - 2;
            ASSERT_EQ(a(x, y), correct_a) << "result(" << x << ", " << y << ") a component incorrect";
            ASSERT_EQ(b(x, y), correct_b) << "result(" << x << ", " << y << ") b component incorrect";
        }
    }
}

TEST(TupleSelect, TernaryWithTupleCondition) {
    // ternary select with Tuple condition
    Var x("x"), y("y");
    Func f("f");

    f(x, y) = select(Tuple(x < 30, y < 30), Tuple(x, y), Tuple(x - 1, y - 2));

    Realization result = f.realize({200, 200});
    Buffer<int> a = result[0], b = result[1];
    for (int y = 0; y < a.height(); y++) {
        for (int x = 0; x < a.width(); x++) {
            int correct_a = (x < 30) ? x : x - 1;
            int correct_b = (y < 30) ? y : y - 2;
            ASSERT_EQ(a(x, y), correct_a) << "result(" << x << ", " << y << ") a component incorrect";
            ASSERT_EQ(b(x, y), correct_b) << "result(" << x << ", " << y << ") b component incorrect";
        }
    }
}

TEST(TupleSelect, MultiwayWithExprCondition) {
    // multiway select with Expr condition
    Var x("x"), y("y");
    Func f("f");

    f(x, y) = select(x + y < 30, Tuple(x, y),
                     x + y < 100, Tuple(x - 1, y - 2),
                     Tuple(x - 100, y - 200));

    Realization result = f.realize({200, 200});
    Buffer<int> a = result[0], b = result[1];
    for (int y = 0; y < a.height(); y++) {
        for (int x = 0; x < a.width(); x++) {
            int correct_a = (x + y < 30) ? x : ((x + y < 100) ? x - 1 : x - 100);
            int correct_b = (x + y < 30) ? y : ((x + y < 100) ? y - 2 : y - 200);
            ASSERT_EQ(a(x, y), correct_a) << "result(" << x << ", " << y << ") a component incorrect";
            ASSERT_EQ(b(x, y), correct_b) << "result(" << x << ", " << y << ") b component incorrect";
        }
    }
}

TEST(TupleSelect, MultiwayWithTupleCondition) {
    // multiway select with Tuple condition
    Var x("x"), y("y");
    Func f("f");

    f(x, y) = select(Tuple(x < 30, y < 30), Tuple(x, y),
                     Tuple(x < 100, y < 100), Tuple(x - 1, y - 2),
                     Tuple(x - 100, y - 200));

    Realization result = f.realize({200, 200});
    Buffer<int> a = result[0], b = result[1];
    for (int y = 0; y < a.height(); y++) {
        for (int x = 0; x < a.width(); x++) {
            int correct_a = (x < 30) ? x : ((x < 100) ? x - 1 : x - 100);
            int correct_b = (y < 30) ? y : ((y < 100) ? y - 2 : y - 200);
            ASSERT_EQ(a(x, y), correct_a) << "result(" << x << ", " << y << ") a component incorrect";
            ASSERT_EQ(b(x, y), correct_b) << "result(" << x << ", " << y << ") b component incorrect";
        }
    }
}
